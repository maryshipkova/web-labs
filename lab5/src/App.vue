<template>
    <div id="app">
        <header class="btn-group">
            <button class="btn">save</button>
            <button class="btn">load</button>
        </header>
        <textarea v-model="input"></textarea>
        <div v-html="compiledMarkdown"></div>
    </div>
</template>

<script>

    export default {
        name: 'app',
        data: function () {
            return {
                input: '# hello'
            };
        },
        computed: {
            compiledMarkdown: function () {
                // eslint-disable-next-line
                return marked(this.input, {sanitize: true})
            }
        },

    }
</script>

<style>
    html, body, #app {
        margin: 0;
        height: 100%;
        font-family: 'Helvetica Neue', Arial, sans-serif;
        color: #333;
    }

    textarea, #app div {
        display: inline-block;
        width: 49%;
        height: 100%;
        vertical-align: top;
        box-sizing: border-box;
        padding: 0 20px;
    }

    textarea {
        outline: none;
        background-color: #f6f6f6;
        font-size: 14px;
        font-family: 'Monaco', courier, monospace;
        padding: 20px;
    }

    code {
        color: #f66;
    }

    .btn-group {
        padding: .5rem;
        background-color: blueviolet;
        /*position: absolute;*/
    }
    .btn{
        padding: .5rem 1rem;
        border-radius: 50%;
    }
    .btn:first-of-type{
        margin-right: 1rem;
    }
</style>
