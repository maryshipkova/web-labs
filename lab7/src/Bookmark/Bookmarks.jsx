import React from 'react';
import './Bookmarks.scss'
import {connect} from "react-redux";
import {add} from '../store/actions.js';

const BookmarksFC = (props) => {
    const [value, setValue] = React.useState('');
    const handleSubmit = (e) => {
        e.preventDefault();
        props.add(value);
    };

    // избранное в localStorage
    return (
        <main className={'Bookmarks'}>
            <div className={'Bookmarks-Row'}>
                <h3 className={'Bookmarks-Title'}>Избранное</h3>
                <form onSubmit={handleSubmit} className={'Bookmarks-Add'}>
                    <input
                        className={'Bookmarks-Input'}
                        type="text"
                        placeholder={'Добавить новый город'}
                        onChange={e => setValue(e.target.value)}
                    />
                    <input type={"submit"} className={'Bookmarks-AddButton'} value={'+'}/>
                </form>
            </div>
        </main>
    )
};


const mapDispatchToProps = {
    add
};

export const Bookmarks = connect(null, mapDispatchToProps)(BookmarksFC);
