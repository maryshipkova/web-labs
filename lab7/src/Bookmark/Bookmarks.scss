@import '../constants.scss';

.Bookmarks {
  &-Row{
    display: flex;
    align-items: center;
    justify-content: space-between;
  }

  &-Input{

  }

  &-AddButton{
    margin-left: 10px;
  }

}
