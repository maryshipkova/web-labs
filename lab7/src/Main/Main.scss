@import '../constants.scss';

.Main {
  display: flex;
  flex-direction: column;

  background-color: white;
  border-radius: 18px;
  min-width: 50%;
  min-height: 200px;
  padding: 10px 20px;

  &-Button{
    background-color:$color-gray;
    padding: 10px 15px;
    border-radius: 100px;
    border: none;
    font-size: 16px;
    font-weight: bold;
  }
}
