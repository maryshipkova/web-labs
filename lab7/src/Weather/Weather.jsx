import * as React from 'react';
import './Weather.scss';

const getData = (city, coordinates) => {
    const fetchUrl = city ? `q=${city}` : `lat=${coordinates.lat}&lon=${coordinates.lon}`;

    return fetch(`http://api.openweathermap.org/data/2.5/forecast?${fetchUrl}&APPID=62462c93c650ac75e405b900f2457d73`)
        .then(data => data.json())
};

export const Weather = (props) => {
    const [weatherInfo, updateWeatherInfo] = React.useState({});
    const [loading, isLoading] = React.useState(false);

    const getCityData = (city) => {
        return getData(city, coordinates)
            .then(data => {
                const currWeather = data.list[0];
                updateWeatherInfo({
                    cityName: data.city.name,
                    coord: [data.city.coord.lat, data.city.coord.lon],
                    humidity: `${currWeather.main.humidity} %`,
                    pressure: `${currWeather.main.pressure} hPa`,
                    wind: `${currWeather.wind.speed} meter/sec`,
                    weather: currWeather.weather[0].description,
                    temperature: `${parseInt((currWeather.main.temp - 272) + '')} C`,
                    icon: currWeather.weather[0].icon
                });
                isLoading(false);
            })
            .catch(e => {
                isLoading(false);
                return updateWeatherInfo({city, error: '404 city not found'});
            })
    };

    const {city, coordinates, main, onRemove} = props;

    React.useEffect(() => {
        if (city || coordinates) {
            isLoading(true);
            getCityData(city, coordinates);
        }
    }, [city, coordinates]);


    if(loading) return <div>loading</div>;
    if (weatherInfo.error){
        return (
            <div>
                {city}: {weatherInfo.error}
                <button className={'Weather-Remove'} onClick={() => onRemove(city)}>&Chi;</button>
            </div>);
    }

    const {cityName, temperature, wind, weather, pressure, humidity, coord, icon} = weatherInfo;

    return (
        <div className={`Weather${main ? ' Weather_main' : ''}`}>
            {weatherInfo && weatherInfo.cityName ?
                <React.Fragment>
                    <div className="Weather-General">
                        <h1 className="Weather-City">{cityName}</h1>
                        <div className="Weather-MainInfo">
                            <div className="Weather-Icon">
                                <img src={`http://openweathermap.org/img/wn/${icon}@2x.png`} alt={icon}/>
                            </div>
                            <h2 className="Weather-Temperature">{temperature}</h2>
                        </div>
                        {!main && <button className={'Weather-Button'} onClick={() => onRemove(city)}>&Chi;</button>}
                    </div>
                    <ul className="Weather-List">
                        <li className="Weather-Item">Ветер: {wind}</li>
                        <li className="Weather-Item">Облачность: {weather}</li>
                        <li className="Weather-Item">Давление: {pressure}</li>
                        <li className="Weather-Item">Влажность: {humidity}</li>
                        <li className="Weather-Item">Координаты: {coord && `${coord[0]} ${coord[1]}`}</li>
                    </ul>
                </React.Fragment> : null}
        </div>
    )
};
