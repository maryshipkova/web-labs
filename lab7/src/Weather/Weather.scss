@import '../constants.scss';

.Weather {
  &:not(:last-child){
    margin-bottom: 20px;
  }

  &-General {
    display: flex;
    align-items: center;
    font-size: 16px;
  }

  &-City {
    margin: 0;
  }

  &-MainInfo {
    display: flex;
    align-items: center;
    margin-left: 10px;
  }

  &-Temperature {
    margin: 0;
    margin-left: 10px;
  }

  &-List {
    padding: 0;
  }

  &-Item {
    padding: 5px 10px;
    margin-bottom: 5px;

    background-color: $color-gray;
    color: white;
    list-style: none;
  }
}

.Weather_main.Weather {
  display: flex;
  margin-top: 20px;

  .Weather-General {
    flex-direction: column;
    align-items: flex-start;
  }

  .Weather-MainInfo{
    margin-top: 20px;

  }

  .Weather-List {
    padding: 0;
    margin-left: 90px;
  }

  .Weather-Temperature {
    font-size: 50px;
  }
}
