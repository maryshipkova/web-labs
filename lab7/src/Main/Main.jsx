import React from 'react';
import {Weather} from "../Weather/Weather";
import {Bookmarks} from "../Bookmarks/Bookmarks";
import {useDispatch, useSelector} from "react-redux";
import './Main.scss'
import {remove} from "../store/actions";

export const Main = (props) => {
    const [coordinates, updateCoordinates] = React.useState();
    const [loading, isLoading] = React.useState(false);
    const cities = useSelector(state => state.cities);
    const dispatch = useDispatch();
    const onButtonClick = () => {
        isLoading(true);

        navigator.geolocation.getCurrentPosition(function (position) {
                const latitude = position.coords.latitude;
                const longitude = position.coords.longitude;
                updateCoordinates({lat: latitude, lon: longitude});
                isLoading(false);
            },
            function (err) {
                console.log(err);
                updateCoordinates({lat: '59', lon: '30'});
                isLoading(false);
            }, {
                maximumAge:1
            });

    };

    const onRemove = (city) => {
        remove(dispatch, city);
    };

    return (
        <main className={'Main'}>
            <div className="Main-Search">
                <h3>Погода здесь</h3>
                <button className="Main-Button" onClick={onButtonClick}>Обновить геолокацию</button>
            </div>
            {
                props.loading || loading ? <div>loading</div> :
                    <Weather main coordinates={coordinates}/>
            }
            <div className={'Main-Bookmarks'}>
                <Bookmarks/>
                {
                    cities && cities.map(city => <Weather id={city} city={city} key={city} onRemove={onRemove}/>)
                }
            </div>
        </main>
    )
};
