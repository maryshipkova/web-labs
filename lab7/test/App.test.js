import React from 'react';
import {Bookmarks} from "../src/Bookmarks/Bookmarks";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import {Weather} from "../src/Weather/Weather";
import {Main} from "../src/Main/Main";
import App from "../src/App";

it('App', () => {
    const initialState={cities: ["Moscow","1234","12345"]};
    const mockStore = configureStore([thunk]);
    const store = mockStore(initialState);

    const tree = renderer
        .create(
            <Provider store={store}>
                <App/>
            </Provider>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
