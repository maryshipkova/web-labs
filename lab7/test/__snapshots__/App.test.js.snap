// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 1`] = `
<div
  className="App"
>
  <main
    className="Main"
  >
    <div
      className="Main-Search"
    >
      <h3>
        Погода здесь
      </h3>
      <button
        className="Main-Button"
        onClick={[Function]}
      >
        Обновить геолокацию
      </button>
    </div>
    <div
      className="Weather Weather_main"
    />
    <div
      className="Main-Bookmarks"
    >
      <main
        className="Bookmarks"
      >
        <div
          className="Bookmarks-Row"
        >
          <h3
            className="Bookmarks-Title"
          >
            Избранное
          </h3>
          <form
            className="Bookmarks-Add"
            onSubmit={[Function]}
          >
            <input
              className="Bookmarks-Input"
              onChange={[Function]}
              placeholder="Добавить новый город"
              type="text"
            />
            <input
              className="Bookmarks-AddButton Weather-Button"
              type="submit"
              value="+"
            />
          </form>
        </div>
      </main>
      <div
        className="Weather"
      />
      <div
        className="Weather"
      />
      <div
        className="Weather"
      />
    </div>
  </main>
</div>
`;
